web/src/peer_data.ts

```diff
@@ -1,15 +1,57 @@
+import {z} from "zod";
+
 import * as blueslip from "./blueslip.ts";
+import * as channel from "./channel.ts";
 import {LazySet} from "./lazy_set.ts";
+import {page_params} from "./page_params.ts";
 import type {User} from "./people.ts";
 import * as people from "./people.ts";
 import * as sub_store from "./sub_store.ts";
 
 // This maps a stream_id to a LazySet of user_ids who are subscribed.
 const stream_subscribers = new Map<number, LazySet>();
 const fetched_stream_ids = new Set<number>();
+export function has_full_subscriber_data(stream_id: number): boolean {
+    return fetched_stream_ids.has(stream_id);
+}
+const pending_subscriber_requests = new Map<number, Promise<LazySet>>();
 
 export function clear_for_testing(): void {
     stream_subscribers.clear();
+    fetched_stream_ids.clear();
+}
+
+const fetch_stream_subscribers_response_schema = z.object({
+    subscribers: z.array(z.number()),
+});
+
+export async function maybe_fetch_stream_subscribers(stream_id: number): Promise<LazySet> {
+    if (pending_subscriber_requests.has(stream_id)) {
+        return pending_subscriber_requests.get(stream_id)!;
+    }
+    const promise = (async () => {
+        let subscribers: number[];
+        try {
+            const xhr = await channel.get({
+                url: `/json/streams/${stream_id}/members`,
+            });
+            subscribers = fetch_stream_subscribers_response_schema.parse(xhr).subscribers;
+        } catch {
+            blueslip.error("Failure fetching channel subscribers", {
+                stream_id,
+            });
+            pending_subscriber_requests.delete(stream_id);
+            // Fall back to what we already have.
+            return get_user_set(stream_id);
```

Comment:
I think this probably want to do some sort of retry with exponential backoff, rather than just giving up and returning a bogus answer? Not sure.

Reply:
I tried `null` because I feel like it's good to be really explicit that we're deliberately not having data here. Happy to do `undefined` too though, if you prefer it.

It's a bit tricky because the call chain is pretty long here, so many functions now maybe return `null`. I put it in a commit and you can see what you think. I wasn't sure what to do in some of the situations (return early or go ahead and display links or empty messages) and I generally chose to just return early, but let me know if you'd prefer something else.

Reply:
That seems OK as a follow-up, yeah.